staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,570 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+)
+
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+// The provided schema is expected to exactly match the fields we expose to CEL. I.e.
+// for resource roots, type meta and object meta schemas must be present.
+func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+	if unstructured == nil {
+		if schema.Nullable {
+			return types.NullValue
+		}
+		return types.NewErr("invalid data, got null for schema with nullable=false")
+	}
+	if schema.XIntOrString {
+		switch unstructured.(type) {
+		case string, int64:
+			return types.DefaultTypeAdapter.NativeToValue(unstructured)
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
+	}
+	if schema.Type == "object" {
+		m, ok := unstructured.(map[string]interface{})
+		if !ok {
+			types.NewErr("invalid data, expected map[string]interface{} to match the provided schema with type=object")
```

Comment:
missing a return? also missing a test case then?

```suggestion
			return types.NewErr("invalid data, expected map[string]interface{} to match the provided schema with type=object")
```